### PR TITLE
feat(cli,app): upload pasted images to cloud storage instead of local .assets

### PR DESCRIPTION
**Pull request by jaden-cis**
## Summary
- `inplan asset-upload`: uploads image bytes straight to the `doc-images` bucket for a cloud-linked doc, mirroring the cloud web app's `saveAsset` (same bucket, org/doc path scheme, 409-collision retry).
- The desktop app's `asset:save` IPC handler routes through it whenever the doc is live-connected to the cloud (Collaborate on Cloud), so a pasted image never needs a later migration.
- `inplan upload` (promote) now migrates any images pasted while the doc was still local: it scans the body for local relative image links, uploads the referenced files, and rewrites the links in both the local file and the cloud row — otherwise a doc promoted after the fact arrives in the cloud with links into a `.assets/` folder that has no counterpart there.

## Test plan
- [x] `packages/cli/test/assetUpload.test.ts` — 8 cases (upload, extension fallback, 409-collision retry, hard failure, logged-out, non-cloud doc, org lookup failure)
- [x] `packages/cli/test/uploadImageMigration.test.ts` — 5 cases (migrates a local image, no-op with no images, leaves absolute URLs alone, leaves a dangling link alone, a storage failure doesn't fail the promote itself)
- [x] Verified end-to-end against the real cloud (Supabase Storage) with both single- and multi-image docs — uploaded images are publicly reachable and the rewritten links match the stored document body.
- [x] `npm run typecheck -w @inplan/cli -w @inplan/app` clean.

<a href="https://cubic.dev/pr/melly-lgtm/inplan/pull/91?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added cloud image uploads for documents connected to the cloud.
  * Local image references are migrated to public URLs during document uploads.
  * Added support for common image formats, with PNG fallback for unknown extensions.
  * Added collision handling and cleanup for temporary upload files.

* **Bug Fixes**
  * Uploads continue safely when individual image migrations fail.
  * Local documents retain their existing asset-storage behavior.
  * Image paste failures now display an error without changing the document.

* **Tests**
  * Expanded coverage for uploads, authentication, failures, collisions, and URL rewriting.

### CODE REVIEW

**Comment by jaden-cis** [comment]
@coderabbitai full review

**Comment by melly-lgtm** [comment]
@cubic review

File: packages/cli/src/cli.ts [hunk 1]
```diff
@@ -1928,17 +1930,159 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+}
+
+/** A Markdown image reference: `![alt](<dest>)` (the angle-bracket form the editor writes for
+ *  asset:save — needed because `<docname>.assets/…` can contain spaces) or the bare `![alt](dest)`
+ *  form (no unescaped whitespace/parens, which the bracket-less destination syntax can't carry). */
+const IMAGE_REF_RE = /!\[([^\]]*)\]\(\s*(?:<([^>]+)>|([^()\s]+))\s*\)/g;
+
+/** True for a relative path (not a URL) — mirrors the renderer's image-src resolver
+ *  (`markdown.ts`) so promote-time migration and live rendering agree on what counts as "local". */
+function isLocalRelativePath(dest: string): boolean {
```
Comment (melly-lgtm) [inline packages/cli/src/cli.ts:2130]: **Security (blocking): path traversal → exfiltration to a public bucket.**

This accepts any relative path, including `../`. Combined with `resolve(docDir, dest)` below, a body link like `` or `` resolves to a file anywhere the user can read — and `migrateLocalImages` reads it and uploads it to the **public** `doc-images` bucket, then rewrites the doc to the public URL.

Doc bodies are not trusted-solely-authored input: plans come from cloned repos, collaborators, and agents write into them. A poisoned image link + a routine `inplan upload` silently publishes arbitrary local files.

Fix (defense in depth): reject anything whose real path escapes the doc dir — `realpathSync(abs)` must start with `realpathSync(docDir) + sep` (realpath also closes the symlink-escape variant), and must be a regular file. Please add a test: `` → untouched, zero uploads.

File: packages/cli/src/cli.ts [hunk 2]
```diff
@@ -1928,17 +1930,159 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+    if (!dest || !isLocalRelativePath(dest) || replacements.has(dest)) continue;
+    const abs = resolve(docDir, dest);
+    if (!existsSync(abs)) continue; // already dangling — nothing to migrate
+    let bytes: Buffer;
+    try {
+      bytes = readFileSync(abs);
+    } catch {
+      continue;
+    }
+    const ext = extname(abs).slice(1).toLowerCase() || "png";
```
Comment (melly-lgtm) [inline packages/cli/src/cli.ts]: **Security (blocking): the png fallback makes the extension check a no-op.**

Falling back to png for any extension means a non-image file (``) is happily read and uploaded declared as `image/png`. The bucket's `allowed_mime_types` won't catch it — Storage validates the *declared* Content-Type, not the bytes.

The png fallback is correct for a live paste (`asset:save`), but wrong here: for a filesystem scan, an unrecognized extension should be **skipped**, not coerced. Only migrate `dest` whose extension is a key of `ASSET_MIME_BY_EXT`.

File: packages/cli/src/cli.ts [hunk 3]
```diff
@@ -1928,17 +1930,159 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+    const url = await uploadAssetBytes(db, orgId, docId, bytes, ext);
+    if (url) replacements.set(dest, url);
+  }
+  if (replacements.size === 0) return { body, migrated: 0 };
+  const newBody = body.replace(IMAGE_REF_RE, (whole, _alt: string, bracketed?: string, bare?: string) => {
```
Comment (melly-lgtm) [inline packages/cli/src/cli.ts]: **Correctness (blocking): the rewrite can hit parenthesized alt text instead of the destination.**

`whole.replace(/\(\s*(?:<[^>]+>|[^()\s]+)\s*\)/, ...)` replaces the *first* parenthesized run in the whole match, which may be inside the alt text. Reproduced:

```
IN : ![a (x) b](img.png)   (img.png migrated)
OUT: ![a (https://cdn/…/X.png) b](img.png)
```

The alt gets a URL, the real link stays local and unmigrated — corrupted and still dangling.

The outer regex already captured alt + destination; rebuild from those instead of re-matching:
```js
return url ? `![${_alt}](${url})` : whole;
```
Add a test with parens in the alt text.

File: packages/cli/src/cli.ts [hunk 4]
```diff
@@ -1928,17 +1930,159 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+  // local⇄cloud reconcile check honest. Best-effort: a migration failure must not fail the promote
+  // itself — the doc is already created; the human can always fall back to re-pasting the images
+  // once collaborating in the cloud.
+  let finalBody = body;
+  try {
+    const migrated = await migrateLocalImages(body, dirname(file), s.db, pick.org_id, cloudDocId);
+    if (migrated.migrated > 0) {
+      finalBody = migrated.body;
+      writeFileSync(file, finalBody);
+      await new SupabaseDocumentStore(s.db, cloudDocId).saveDoc(finalBody);
```
Comment (melly-lgtm) [inline packages/cli/src/cli.ts]: **Architecture (blocking): second writer to `documents.body`.**

Under the unified-Yjs model the collab hub is the **sole** writer of `documents.body` — it's a projection serialized by `onStoreDocument` (`inplan-cloud packages/collab/src/unifiedDoc.ts`, `server.ts`). The old dual-writer was removed deliberately after body corruption. This `saveDoc` reintroduces one.

For `status:"created"` it's mostly tolerable (no Yjs session exists yet), though there's a small race if a web user opens the doc between the RPC and this write. For the `status:"exists"` re-upload branch it's not: that doc can be live in the web right now, and the hub's next projection silently overwrites this body write — the cloud copy loses the migrated URLs while the local file keeps them, and `lastSyncedHash` (hash of `finalBody`) no longer matches the cloud body, confusing the next reconcile.

Minimum: run the migration + `saveDoc` only when `out0.status === "created"`. If exists-path migration is needed, route it through the hub the way gated edits go.

Separately, ordering: you rewrite the local file before `saveDoc`. If `saveDoc` throws, `lastSyncedHash` matches the new local body but the cloud keeps the old one, so reconcile pulls the cloud copy back and undoes the local rewrite. Cloud-first, then local, then hash fails cleaner.

**Review by melly-lgtm — CHANGES_REQUESTED** [review]
Solid direction, and the mirroring of the web's `saveAsset` is faithful — same bucket, path scheme, hasOwnProperty guard, and 409 retry (verified line-for-line, and the `error.status` numeric check matches the bundled storage-js `StorageApiError` shape). The RLS note is accurate too: `doc_images_insert` checks the path's org segment against `documents.org_id` AND `app_doc_role in ('owner','editor')`, so a wrong-org upload fails server-side.

Three things block merge — one security, one correctness, one architectural. Details inline.

**1. Security (migration reads/publishes arbitrary local files).** `isLocalRelativePath` accepts `../`, and the extension check falls back to png for anything — so `` in a doc body gets read and uploaded to the *public* bucket declared as image/png (the bucket's mime allowlist validates the declared Content-Type, not bytes), with the doc rewritten to advertise the public URL. Doc bodies aren't trusted-solely-authored input (cloned repos, collaborators, agents write them). Fix: in the migration path accept only extensions in `ASSET_MIME_BY_EXT` (skip, don't png-fallback), and containment-check `realpathSync(abs)` is a regular file under `realpathSync(docDir)+sep`.

**2. Correctness (rewrite can corrupt links with parenthesized alt text).** `a (x) b` rewrites the URL into the alt-text parens and leaves the real destination local. Rebuild from the captured groups instead of doing regex surgery on `whole`.

**3. Architecture (second writer to `documents.body`).** Under the unified-Yjs model the collab hub is the *sole* writer of `documents.body` (`onStoreDocument`); the dual-writer was removed deliberately. The re-upload (`status:"exists"`) branch can write a body the hub then clobbers. Gate the migration to `status:"created"` only, or route the change through the hub.

Everything else — tests (non-vacuous), version skew (uses the bundled CLI), cross-repo duplication (license-forced, both sides carry pointers) — is fine.

**Comment by jaden-cis** [comment]
@coderabbitai full review

File: packages/cli/src/cli.ts (hunk 1 above)
Comment (jaden-cis) [reply packages/cli/src/cli.ts:2130]: Fixed in [a91c212](https://github.com/melly-lgtm/inplan/commit/a91c212902875a75581ee61bbd9184d5deb98c67) — `migrateLocalImages` now containment-checks `realpathSync(abs)` against `realpathSync(docDir)` (via `relative()`, rejecting anything starting with `..` or absolute), so a path-traversal link is skipped entirely, never read or uploaded. Added a test covering exactly this (`../` escaping the doc dir → zero uploads, link left untouched).

File: packages/cli/src/cli.ts (hunk 2 above)
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in [a91c212](https://github.com/melly-lgtm/inplan/commit/a91c212902875a75581ee61bbd9184d5deb98c67) — the migration path no longer falls back to png for an unrecognized extension; it now only migrates `dest` whose extension is a key of `ASSET_MIME_BY_EXT`, skipping anything else. Added a test (a non-image extension → zero uploads, link left untouched).

File: packages/cli/src/cli.ts (hunk 3 above)
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in [c7a337f](https://github.com/melly-lgtm/inplan/commit/c7a337f518d041696de00b4319fdfb05518ad17f) — using exactly the rebuild you suggested: `return url ? `${alt}` : whole;` instead of regex surgery on `whole`. Added a test with parens in the alt text (`a (x) b`) asserting the alt text is preserved verbatim and the destination is the one rewritten.

File: packages/cli/src/cli.ts [hunk 5]
```diff
@@ -1928,17 +1930,159 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+ * command genuinely would be pointless, since nothing would ever call it for the images that
+ * matter most (the ones already in the doc when it's promoted). Best-effort per image: a missing
+ * file or a failed upload just leaves that one link untouched rather than losing the reference.
+ */
+async function migrateLocalImages(body: string, docDir: string, db: SupabaseClient, orgId: string, docId: string): Promise<{ body: string; migrated: number }> {
+  const replacements = new Map<string, string>(); // matched destination → new public URL
+  for (const m of body.matchAll(IMAGE_REF_RE)) {
+    const dest = m[2] ?? m[3] ?? "";
+    if (!dest || !isLocalRelativePath(dest) || replacements.has(dest)) continue;
+    const abs = resolve(docDir, dest);
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts:2215]: Confirmed, thanks — fixed in a91c212 (containment check + extension allowlist).

File: packages/cli/src/cli.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in c7a337f — migration + the cloud saveDoc now only run when `out0.status === "created"`; the exists/re-upload path never writes documents.body.

Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Partially fixed in c7a337f: migration + `saveDoc` now only run when `out0.status === "created"`, so the exists/re-upload branch never writes documents.body — the main dual-writer concern for a live/collaborated doc is gone.

The ordering point (cloud-first, then local, then hash) is still open — currently it's local-write → cloud-write → hash, so a `saveDoc` failure after a successful local write leaves `lastSyncedHash` claiming sync when the cloud still has the old body. Tracking this together with cubic's identical finding on the same line; will follow up separately.

File: packages/cli/src/cli.ts [hunk 6]
```diff
@@ -1928,17 +1930,159 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+  // once collaborating in the cloud.
+  let finalBody = body;
+  try {
+    const migrated = await migrateLocalImages(body, dirname(file), s.db, pick.org_id, cloudDocId);
+    if (migrated.migrated > 0) {
+      finalBody = migrated.body;
+      writeFileSync(file, finalBody);
+      await new SupabaseDocumentStore(s.db, cloudDocId).saveDoc(finalBody);
+    }
+  } catch (e) {
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Valid, not fixed yet — same underlying issue melly-lgtm flagged (the ordering point on the architecture thread): local file is written before the cloud `saveDoc`, so a failure there leaves `lastSyncedHash` matching the new local body while the cloud still has the old one. Will fix by writing cloud first, then local, then hashing — following up separately from the 3 issues just addressed.

File: packages/cli/src/cli.ts (hunk 4 above)
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Now fully fixed in 1779b62 — reordered to cloud-write → local-write → assign finalBody, so a saveDoc failure leaves both the local file and lastSyncedHash at the original (pre-migration) body. Added a test asserting the local file is untouched when the cloud write fails.

File: packages/cli/src/cli.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in 1779b62 — same fix as the architecture thread's ordering point (cloud-first, then local, then assign finalBody only after both succeed).

File: packages/app/src/main/index.ts [hunk 7]
```diff
@@ -116,6 +118,36 @@ function runCli(args: string[], extraEnv?: Record<string, string>): Promise<{ co
...
+ *  directly, instead of writing them to a local `.assets/` folder that would need migrating
+ *  later. The bytes cross the process boundary via a temp file (IPC args are JSON-only) — this
+ *  function owns that file's whole lifecycle, cleaning it up whether the upload succeeds or not.
+ *  Returns null on any failure (network, auth, upload) so the caller falls back the same way a
+ *  local write failure does. */
+async function uploadAssetToCloud(docFile: string, bytes: ArrayBuffer, ext: string): Promise<{ relPath: string } | null> {
+  const tmp = join(tmpdir(), `inplan-asset-${randomUUID()}`);
+  try {
+    writeFileSync(tmp, Buffer.from(bytes));
+    const r = await runCli(["asset-upload", docFile, "--bytes-file", tmp, "--ext", ext]);
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/app/src/main/index.ts]: Fixed in edf26e4 — `writeFileSync(tmp, ..., { mode: 0o600 })`.

File: packages/cli/src/cli.ts [hunk 8]
```diff
@@ -1928,17 +1930,159 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+  const contentType = ASSET_MIME_BY_EXT[safeExt]!;
+  const stamp = new Date().toISOString().replace(/[^0-9]/g, "").slice(0, 14); // YYYYMMDDHHMMSS
+  for (let n = 0; n <= 5; n++) {
+    const name = n === 0 ? `image-${stamp}.${safeExt}` : `image-${stamp}-${n}.${safeExt}`;
+    const path = `${orgId}/${docId}/${name}`;
+    const { error } = await db.storage.from(ASSET_BUCKET).upload(path, bytes, { contentType });
+    if (!error) return db.storage.from(ASSET_BUCKET).getPublicUrl(path).data.publicUrl;
+    if (error.status !== 409) return null; // not a name collision — a real failure, don't retry
+  }
+  return null; // exhausted collision retries
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts:2064]: Fixed in d557379 — every upload attempt now gets its own random 4-byte hex suffix instead of relying on the n<=5 counter for uniqueness, so a same-second batch (migrateLocalImages) never depends on the retry loop. Added a test migrating 7 same-second images.

**Comment by melly-lgtm** [comment]
@cubic review

File: packages/cli/src/cli.ts [hunk 9]
```diff
@@ -1928,17 +1930,213 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+      const migrated = await migrateLocalImages(body, dirname(file), s.db, pick.org_id, cloudDocId);
+      if (migrated.migrated > 0) {
+        // Cloud first, then local, then only NOW does `finalBody` (and so lastSyncedHash below)
+        // reflect the migrated body. If saveDoc throws, we never reach the local write or the
+        // reassignment — the local file and finalBody both stay at the ORIGINAL body, so the
+        // persisted hash always matches whatever the local file actually contains. The old
+        // local-first order could leave lastSyncedHash claiming sync (hash of the migrated body)
+        // while the cloud still held the pre-migration one, on nothing more than a failed cloud
+        // write (flagged in review: cubic-dev-ai + melly-lgtm on PR #91).
+        await new SupabaseDocumentStore(s.db, cloudDocId).saveDoc(migrated.body);
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in 5eb4e9d — the promote-time cloud write is now a conditional UPDATE (`.eq("updated_at", baseline)`), where baseline is the row's `updated_at` read right after create_document. If the hub (or anyone else) touched the row in the meantime, the CAS condition matches zero rows and we don't overwrite it — no schema change needed, updated_at already exists.

File: packages/app/src/main/index.ts [hunk 10]
```diff
@@ -116,6 +118,38 @@ function runCli(args: string[], extraEnv?: Record<string, string>): Promise<{ co
...
+ *  function owns that file's whole lifecycle, cleaning it up whether the upload succeeds or not.
+ *  Returns null on any failure (network, auth, upload) so the caller falls back the same way a
+ *  local write failure does. */
+async function uploadAssetToCloud(docFile: string, bytes: ArrayBuffer, ext: string): Promise<{ relPath: string } | null> {
+  const tmp = join(tmpdir(), `inplan-asset-${randomUUID()}`);
+  try {
+    // 0o600: the OS temp dir is shared system-wide — a permissive umask-derived mode would let
+    // another local user read the pasted image bytes before cleanup.
+    writeFileSync(tmp, Buffer.from(bytes), { mode: 0o600 });
+    const r = await runCli(["asset-upload", docFile, "--bytes-file", tmp, "--ext", ext]);
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/app/src/main/index.ts]: Fixed in 6ac4efb — runCli now accepts an optional timeoutMs; the asset-upload call passes 30s, so a stall falls through to the existing null-return failure path instead of hanging.

File: packages/cli/src/cli.ts [hunk 11]
```diff
@@ -1928,17 +1930,213 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+  // once, up front — a missing docDir (shouldn't happen; the doc file itself was just read) means
+  // nothing here can be verified as contained, so no image gets migrated rather than trusting an
+  // unresolved path.
+  let docDirReal: string;
+  try {
+    docDirReal = realpathSync(docDir);
+  } catch {
+    return { body, migrated: 0 };
+  }
+  for (const m of body.matchAll(IMAGE_REF_RE)) {
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts:2207]: Fixed in dd6af70 — computes the body's fenced/inline code ranges up front and tracks matches by position through both the scan and rewrite passes, so a code example is skipped for upload and left untouched by the rewrite even when it names the exact same path as a real, migrated image elsewhere in the doc.

File: packages/cli/src/cli.ts [hunk 12]
```diff
@@ -1928,17 +1930,213 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+      if (migrated.migrated > 0) {
+        // Cloud first, then local, then only NOW does `finalBody` (and so lastSyncedHash below)
+        // reflect the migrated body. If saveDoc throws, we never reach the local write or the
+        // reassignment — the local file and finalBody both stay at the ORIGINAL body, so the
+        // persisted hash always matches whatever the local file actually contains. The old
+        // local-first order could leave lastSyncedHash claiming sync (hash of the migrated body)
+        // while the cloud still held the pre-migration one, on nothing more than a failed cloud
+        // write (flagged in review: cubic-dev-ai + melly-lgtm on PR #91).
+        await new SupabaseDocumentStore(s.db, cloudDocId).saveDoc(migrated.body);
+        writeFileSync(file, migrated.body);
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in 5eb4e9d, alongside the CAS fix — the cloud write and the local write are now in separate try/catches with distinct error messages, and a local-write failure re-reads whatever's actually on disk for finalBody instead of assuming the intended content landed, so lastSyncedHash never overstates sync in either direction.

**Comment by jaden-cis** [comment]
Addressed both nitpicks from the latest CodeRabbit review, in ae78699:

- Each of the three migration-failure tests (cloud write fails, lost CAS race, local write fails) now directly asserts `lastSyncedHash` from the persisted `status.json` equals `hashBody(original)`, not just that the local file's content looks right.
- The `update` mock now records every `.eq(column, value)` call; the lost-CAS-race test asserts both the `id` and `updated_at` predicates were actually issued by the production code, not just that the mock's canned "zero rows" result was handled.

**Comment by melly-lgtm** [comment]
@cubic review

**Review by melly-lgtm — COMMENTED** [review]
Reviewed at head `ae78699` — pulled the branch, ran the full CLI suite (373/373 incl. the 24 new asset tests) and typecheck (clean), and read every changed file.

## Verdict
This has become a strong PR. The earlier review round's P1s are genuinely fixed — I verified each against head, not just the replies: traversal is closed with realpath containment **on both sides** plus a destination-decided extension allowlist (no png-guessing before a disk read); the exists/adopt path never writes `body`; `lastSyncedHash` now records what's actually on disk in every failure permutation. And the additions go beyond what review asked: optimistic concurrency on the promote-time body write, code-fence/inline-code exclusion, and the position-keyed rewrite that kills both the alt-text-parens bug and the code-example-same-path bug in one move. The failure-path separation in the migration block (cloud-write failed vs local-write-back failed, each with its own truthful message and hash) is exactly the kind of care this path needs.

Remaining items are minor — none blocking, two worth doing:

### 1. CAS baseline has a small create→read window (optional hardening)
The `updated_at` baseline is read *after* `create_document` returns, so a hub attach + debounced persist landing in that gap becomes the baseline — the migration write would then pass the CAS and overwrite the hub's write. The residual risk is tiny (the row was just seeded from this same body; a human would need to open *and* edit within the gap), but the stronger predicate costs nothing: condition on **content** instead of time — `.eq("body", body)` (the exact original the migration derived from). That's timing-independent, immune to clock skew, and drops the extra baseline read entirely.

### 2. A failed cloud paste should reach the user, not stderr (follow-up is fine)
`asset:save` on a cloud doc returns `null` on upload failure and logs to Electron's stderr. The *decision* not to fall back to a local `.assets/` write is correct — a local fallback would strand exactly the images the adopt path (rightly) never migrates — but the human just sees a paste that silently did nothing. Worth confirming the renderer surfaces the null as a visible "image upload failed" state, and if not, filing the follow-up.

### Nits (no action needed to merge)
- **`updated_at` from the client clock** in the CAS write: skew can move the row backwards relative to hub writes, and `updated_at` feeds LRU-eviction ordering. Fine for now; noting for the record.
- **Indented (4-space) code blocks** aren't excluded by `codeRanges` — only fences and inline spans. Rare in plan docs.
- **`relPath` naming**: on the cloud path it carries an absolute public URL. The shape mirrors the local handler so the renderer contract holds, but the name now lies — a rename (`linkDest`?) next time someone touches it.
- **For the record**: migrated images land at public-but-unguessable URLs (org/doc UUID path), matching the web's `saveAsset` — pre-existing model, no action.

### Verified specifics worth calling out as good
- Temp-file handling in `uploadAssetToCloud`: `0o600` in a shared tmpdir, bounded subprocess (30s), cleanup in `finally` — all three of the classic mistakes avoided.
- Desktop preview renders the new absolute URLs (CSP already carries `img-src https:` — checked).
- `>6 same-second images` collision-budget test; unterminated-fence-to-EOF handling in `codeRanges`; help text + SPDX headers on the new files.

File: packages/cli/src/cli.ts [hunk 13]
```diff
@@ -1928,17 +1929,294 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+        // unconditional write here would race it. Condition on `updated_at` being unchanged
+        // since our baseline; the hub's own writes bump it too (create_document's LRU comment:
+        // "autosaves advance it"), so if it moved, someone else touched the row first and this
+        // must NOT overwrite them (flagged in review: cubic-dev-ai on PR #91). No schema change
+        // needed — `updated_at` already exists; this is a plain conditional UPDATE.
+        const { data: updatedRows, error: updateErr } = await s.db
+          .from("documents")
+          .update({ body: migrated.body, updated_at: new Date().toISOString() })
+          .eq("id", cloudDocId)
+          .eq("updated_at", baselineUpdatedAt)
```
Comment (melly-lgtm) [inline packages/cli/src/cli.ts]: Optional hardening: condition on content instead of time — `.eq("body", body)` (the exact original this migration derived from). Timing-independent, immune to clock precision/skew, and the create→baseline-read gap disappears along with the extra read. If the hub wrote an identical body in that window, overwriting is harmless by definition.

File: packages/app/src/main/index.ts [hunk 14]
```diff
@@ -543,11 +585,18 @@ function registerIpc(): void {
...
+  // cloud (Collaborate on Cloud), it goes straight to the cloud's doc-images bucket, so both
+  // sides see it immediately and no local-only image is ever left needing a later migration.
+  // Otherwise (the local-first default) it's written into a `<docname>.assets/` folder next to
+  // the doc (Typora-style), committed to git alongside it — never the sidecar dir, so the
+  // Markdown link it produces stays valid for anyone who clones the repo.
+  ipcMain.handle("asset:save", async (_e, bytes: ArrayBuffer, ext: string) => {
     if (!session) return null;
+    const status = readStatus(session.paths.statusPath);
+    if (status.location === "cloud" && status.cloudDocId) {
+      return uploadAssetToCloud(session.paths.file, bytes, ext);
```
Comment (melly-lgtm) [inline packages/app/src/main/index.ts:598]: Right call not to fall back to a local `.assets/` write here (it would strand exactly the images the adopt path never migrates). But on `null` the human just sees a paste that did nothing, with the reason only on Electron's stderr — worth confirming the renderer shows a visible "image upload failed" state, or filing that follow-up.

**Review by melly-lgtm — APPROVED** [review]
**Approved, conditional on addressing the two actionable items from my review** (review 4915007074):

1. **CAS on content, not time** — switch the promote-time conditional write's predicate to `.eq("body", body)` (the exact original the migration derived from), closing the create→baseline-read window and dropping the extra `updated_at` read.
2. **Surface a failed cloud paste to the user** — confirm the renderer shows a visible "image upload failed" state when `asset:save` returns `null` on the cloud path; if it doesn't, either wire it up here or file the follow-up issue before merging.

The nits (client-clock `updated_at`, indented code blocks, `relPath` naming) are non-blocking — take them or leave them. Everything else verified at head `ae78699`: full CLI suite 373/373, typecheck clean, traversal/adopt/hash-honesty fixes all confirmed against the code, not the replies.

File: packages/cli/src/cli.ts [hunk 15]
```diff
@@ -1928,17 +1929,294 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+    // and publish it as one anyway.
+    const ext = extname(abs).slice(1).toLowerCase();
+    if (!Object.prototype.hasOwnProperty.call(ASSET_MIME_BY_EXT, ext)) continue;
+    let bytes: Buffer;
+    try {
+      bytes = readFileSync(abs);
+    } catch {
+      continue;
+    }
+    const url = await uploadAssetBytes(db, orgId, docId, bytes, ext);
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in 7549d06: `uploadAssetBytes` now returns `{path, url}`, `migrateLocalImages` tracks every path it actually uploaded, and `doUpload` best-effort deletes them via `storage.remove()` on both the CAS-loss and cloud-write-error branches. Covered by new assertions in the existing "cloud write fails" / "lost the CAS race" tests plus a negative assertion that cleanup is *not* called when only the local write fails (the image is still correctly referenced then).

File: packages/cli/src/cli.ts [hunk 16]
```diff
@@ -1928,17 +1929,294 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+        if (updateErr) throw new Error(`cloud write failed: ${updateErr.message}`);
+        if (!updatedRows || updatedRows.length === 0) {
+          throw new Error("the document changed in the cloud while migrating images (a live session attached first) — not overwriting it; re-run `inplan upload` to retry");
+        }
+        // The cloud write succeeded — a failure past this point is a DIFFERENT failure (the
+        // images and the cloud body are already correct), so it must not be reported as "the
+        // cloud write failed", and finalBody must reflect whatever actually ends up on disk
+        // rather than what we merely intended to write (flagged in review: cubic-dev-ai on #91).
+        try {
+          writeFileSync(file, migrated.body);
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts:1999]: Fixed in 96a3eef: compares the current on-disk content against the `body` snapshot right before the local write, and skips the write (leaving the file untouched for the human to reconcile) if it diverged — the cloud copy already has the correct migrated links in that case, so only the local file needs attention. Covered by a new test that edits the file mid-flight (during the mocked upload call) and asserts the edit survives, not the stale snapshot.

File: packages/cli/src/cli.ts [hunk 17]
```diff
@@ -1928,17 +1929,294 @@ export async function doUpload(file: string, args: string[]): Promise<void> {
...
+        // needed — `updated_at` already exists; this is a plain conditional UPDATE.
+        const { data: updatedRows, error: updateErr } = await s.db
+          .from("documents")
+          .update({ body: migrated.body, updated_at: new Date().toISOString() })
+          .eq("id", cloudDocId)
+          .eq("updated_at", baselineUpdatedAt)
+          .select("id");
+        if (updateErr) throw new Error(`cloud write failed: ${updateErr.message}`);
+        if (!updatedRows || updatedRows.length === 0) {
+          throw new Error("the document changed in the cloud while migrating images (a live session attached first) — not overwriting it; re-run `inplan upload` to retry");
```
[the remark this answers is not in the document]
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in 96a3eef: both messages now describe the actual recovery instead of a retry that doesn't work — the lost-the-race path says the images weren't migrated and to re-paste them once collaborating in the cloud, and the local-write-failure path says the local file is stale and points at opening the doc in the editor (or `inplan wait`) to pull the migrated body back down.

File: packages/cli/src/cli.ts (hunk 13 above)
Comment (jaden-cis) [reply packages/cli/src/cli.ts]: Fixed in e5d2d95: switched to `.eq("body", body)`, the exact original body this migration derived from, and dropped the separate `updated_at` baseline read entirely (no longer needed once the condition is content-based). Closes the create→baseline-read window and drops the extra read as a bonus.

File: packages/app/src/main/index.ts (hunk 14 above)
Comment (jaden-cis) [reply packages/app/src/main/index.ts:598]: Confirmed and fixed in d19abe9: the renderer's `onPickImage` (App.tsx) — the single call site all three paste paths (source-editor paste, preview paste, toolbar file picker) route through — now sets the existing "couldn't paste image" status message whenever `saveAsset` resolves to null, not just when it throws. Previously only the preview-paste handler's catch block showed anything, and only for a thrown exception; a resolved null (the actual shape of a cloud-upload failure) was silently dropped everywhere. Added tests for the resolved-null case in both the preview-paste and source-paste suites.
